schemes.go

```diff
@@ -921,6 +925,71 @@ func (scheme *SchemeManager) parseContents(conf *Configuration) error {
 		conf.Issuers[issuer.Identifier()] = issuer
 		return scheme.parseCredentialsFolder(conf, issuer, filepath.Join(dir, "Issues"))
 	})
+
+	if err != nil {
+		return err
+	}
+
+	// get a map of CredentialTypes of the current scheme
+	allCredTypesForScheme := map[CredentialTypeIdentifier]CredentialType{}
+	for _, credType := range conf.CredentialTypes {
+		if credType.SchemeManagerID == scheme.ID {
+			allCredTypesForScheme[credType.Identifier()] = *credType
+		}
+	}
+
+	// validate that there are no circular dependencies
+	for _, item := range allCredTypesForScheme {
+		if err := item.validateDependencies(conf, []CredentialTypeIdentifier{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ct CredentialType) validateDependencies(conf *Configuration, validatedDeps DependencyChain) error {
+	if len(validatedDeps) >= maxDepComplexity {
+		return errors.New("dependency tree too complex: " + validatedDeps.String())
+	}
+	for _, outer := range conf.CredentialTypes[ct.Identifier()].Dependencies {
+		for _, middle := range outer {
+			for _, item := range middle {
+				if conf.CredentialTypes[item].SchemeManagerID != ct.SchemeManagerID {
+					return errors.New("credential type " + ct.Identifier().String() + " in scheme " + ct.SchemeManagerID +
+						" has dependency outside the scheme: " + conf.CredentialTypes[item].Identifier().String())
+				}
+
+				if ok := validatedDeps.contains(item); ok {
+					return errors.New("circular dependency " + item.String() + " found in: " + validatedDeps.String())
+				}
+
+				if conf.CredentialTypes[item].Dependencies != nil {
+					validatedDeps = append(validatedDeps, item)
+					return conf.CredentialTypes[item].validateDependencies(conf, validatedDeps)
```

Comment:
This will make the function exit as soon as it encounters a credential type that itself has dependencies, even if there are more credentials present in `conf.CredentialTypes[ct.Identifier()].Dependencies`. Did you mean to return only in case of error, instead of always?

Additionally, this seems to ignore the disjunction structure: it enforces that no credential type occurs a second time anywhere, if it is found once anywhere within the `Dependencies`. But if credential type X occurs in disjunction 1, then it would be fine if it also occurs in disjunction 2, or in dependencies of credential types of disjunction 2.

Reply:
Right, the check was not implemented properly. As discussed orally, the dependency check should pass if there is at least one valid dependency flow.